tensorflow/core/kernels/fixed_length_record_reader_op.cc

```diff
@@ -29,73 +32,137 @@ class FixedLengthRecordReader : public ReaderBase {
         header_bytes_(header_bytes),
         record_bytes_(record_bytes),
         footer_bytes_(footer_bytes),
         hop_bytes_(hop_bytes),
         env_(env),
-        file_pos_limit_(-1),
-        record_number_(0) {}
+        record_number_(0),
+        encoding_(encoding) {}
 
   // On success:
-  // * input_buffer_ != nullptr,
-  // * input_buffer_->Tell() == header_bytes_
-  // * file_pos_limit_ == file size - footer_bytes_
+  // * buffered_input_stream_ != nullptr,
+  // * buffered_input_stream_->Tell() == header_bytes_
   Status OnWorkStartedLocked() override {
     record_number_ = 0;
-    uint64 file_size = 0;
-    TF_RETURN_IF_ERROR(env_->GetFileSize(current_work(), &file_size));
-    file_pos_limit_ = file_size - footer_bytes_;
 
-    TF_RETURN_IF_ERROR(env_->NewRandomAccessFile(current_work(), &file_));
+    lookahead_cache_.clear();
 
-    input_buffer_.reset(new io::InputBuffer(file_.get(), kBufferSize));
-    TF_RETURN_IF_ERROR(input_buffer_->SkipNBytes(header_bytes_));
+    TF_RETURN_IF_ERROR(env_->NewRandomAccessFile(current_work(), &file_));
+    if (encoding_ == "ZLIB" || encoding_ == "GZIP") {
+      const io::ZlibCompressionOptions zlib_options =
+          encoding_ == "ZLIB" ?
+          io::ZlibCompressionOptions::DEFAULT() :
+          io::ZlibCompressionOptions::GZIP();
+      file_stream_.reset(new io::RandomAccessInputStream(file_.get()));
+      buffered_input_stream_.reset(new io::ZlibInputStream(file_stream_.get(),
+                                                          (size_t)kBufferSize,
+                                                          (size_t)kBufferSize,
+                                                          zlib_options));
+    } else {
+      buffered_input_stream_.reset(
+          new io::BufferedInputStream(file_.get(), kBufferSize));
+    }
+    TF_RETURN_IF_ERROR(buffered_input_stream_->SkipNBytes(header_bytes_));
+
+    int bytes_to_read = footer_bytes_;
+
+    // In case hop_bytes_ is in between 0 and record_bytes_,
+    // we will need to hold a cache so that later on we could prefix the cache
+    // with the remaining data to read:
+    // For example, assume record_bytes is 2 and hop_bytes is 1,
+    // For file "H123",
+    // We will process the data in the following way:
+    // 1. Read hope_bytes of 1 ("1") and save it in hop cache.
+    // 2. Read record_bytes - hop_bytes = 1 ("2")
+    // 3. Prefix data in 2. with hop cache in one, we have "12",
+    //    this is used for the record.
+    // 4. Shift hop_bytes of 1 from the record, and we put "2" to hop cache.
+    // 5. Continue step 2.
+    // In order to acheive the above, we will peek in "record_bytes_ - hop_bytes_"
+    // before we read the first record. 
+    if (0 < hop_bytes_ && hop_bytes_ < record_bytes_) {
+      bytes_to_read += record_bytes_ - hop_bytes_;
+    }
+    TF_RETURN_IF_ERROR(buffered_input_stream_->ReadNBytes(bytes_to_read, &lookahead_cache_));
```

Comment:
What happens if the file is empty, i.e., has no records? ZlibInputStream would throw a OutOfRangeError here. May be defer filling cache till first call to ReadLocked?
It would be super awesome if you could add a test for this!